doc/neps/nep-0031-uarray.rst

```diff
@@ -0,0 +1,394 @@
+
+And a library that implements a NumPy-like API will use it in the following
+manner (as an example)::
+
+    import numpy.overridable as np
+    _ua_implementations = {}
+
+    __ua_domain__ = "numpy"
+
+    def __ua_function__(func, args, kwargs):
+        fn = _ua_implementations.get(func, None)
+        return fn(*args, **kwargs) if fn is not None else NotImplemented
+
+    def implements(ua_func):
+        def inner(func):
+            _ua_implementations[ua_func] = func
+            return func
+
+        return inner
+
+    @implements(np.asarray)
+    def asarray(a, dtype=None, order=None):
+        # Code here
+        # Either this method or __ua_convert__ must
+        # return NotImplemented for unsupported types,
+        # Or they shouldn't be marked as dispatchable.
+
+    # Provides a default implementation for ones and zeros.
+    @implements(np.full)
+    def full(shape, fill_value, dtype=None, order='C'):
+        # Code here
+
+The only change this NEP proposes at its acceptance, is to make ``unumpy`` the
+officially recommended way to override NumPy. ``unumpy`` will remain a separate
+repository/package (which we propose to vendor to avoid a hard dependency, and
+use the separate ``unumpy`` package only if it is installed) rather than depend
+on for the time being). In concrete terms, ``numpy.overridable`` becomes an
+alias for ``unumpy``, if available with a fallback to the a vendored version if
+not. ``uarray`` and ``unumpy`` and will be developed primarily with the input
+of duck-array authors and secondarily, custom dtype authors, via the usual
+GitHub workflow. There are a few reasons for this:
+
+* Faster iteration in the case of bugs or issues.
+* Faster design changes, in the case of needed functionality.
+* ``unumpy`` will work with older versions of NumPy as well.
+* The user and library author opt-in to the override process,
+  rather than breakages happening when it is least expected.
+  In simple terms, bugs in ``unumpy`` mean that ``numpy`` remains
+  unaffected.
+
+Duck-array coercion
+~~~~~~~~~~~~~~~~~~~
+
+There are inherent problems about returning objects that are not NumPy arrays
+from ``numpy.array`` or ``numpy.asarray``, particularly in the context of C/C++
+or Cython code that may get an object with a different memory layout than the
+one it expects. However, we believe this problem may apply not only to these
+two functions but all functions that return NumPy arrays. For this reason,
+overrides are opt-in for the user, by using the submodule ``numpy.overridable``
+rather than ``numpy``. NumPy will continue to work unaffected by anything in
+``numpy.overridable``.
+
+If the user wishes to obtain a NumPy array, there are two ways of doing it:
+
+1. Use ``numpy.asarray`` (the non-overridable version).
+2. Use ``numpy.overridable.asarray`` with the NumPy backend set and coercion
+   enabled::
+
+    import numpy.overridable as np
+
+    with ua.set_backend(np):
+        x = np.asarray(...)
```

Comment:
I would move this section way up. Because it is the one concrete thing that explains how it makes users lives easier.

It seems to me, as a library author I could write:
```
import unumpy as unp

def library_function(array):
    array = unp.asarray(array)
    # Code using array, maybe even converting it to numpy array using a with statement locally?
    return array

# On the user side:
library_function(dask_array)  # works and returns dask_array

with unp.set_backend("Dask"):
    library_function([1, 2, 3, 4])  # actually returns a Dask array.
```